File: jiseok/LeetCode/295.py
class MedianFinder:

    def __init__(self):
        self.int_list = []

    def addNum(self, num: int) -> None:
        self.int_list.append(num)

    def findMedian(self) -> float:
        # sort가 O(NlogN)이니까 TLE 날 것
        # 다른 방법 생각 안 나네
        self.int_list.sort()
        N = len(self.int_list)
        half = N // 2
        if N % 2 == 1:
            return float(self.int_list[half])
        else:
            return (self.int_list[half - 1] + self.int_list[half]) / 2


# Your MedianFinder object will be instantiated and called as such:
# obj = MedianFinder()
# obj.addNum(num)
# param_2 = obj.findMedian()

File: dong99u/leetcode/295.java
package leetcode.leetcode295;

import java.util.*;

public class MedianFinder {

    PriorityQueue<Integer> lower; // Max-Heap
    PriorityQueue<Integer> upper; // Min-Heap

    public MedianFinder() {
        this.lower = new PriorityQueue<>(Collections.reverseOrder());
        this.upper = new PriorityQueue<>();
    }

    public void addNum(int num) {
        if (lower.isEmpty() || num <= lower.peek()) {
            lower.offer(num);
        } else {
            upper.offer(num);
        }
        // 균형 조정
        if (lower.size() > upper.size() + 1) {
            upper.offer(lower.poll());
        } else if (upper.size() > lower.size()) {
            lower.offer(upper.poll());
        }
    }

    public double findMedian() {
        if (lower.size() == upper.size()) {
            return (lower.peek() + upper.peek()) / 2.0;
        }
        return lower.peek();
    }
}

/**
 * Your MedianFinder object will be instantiated and called as such:
 * MedianFinder obj = new MedianFinder();
 * obj.addNum(num);
 * double param_2 = obj.findMedian();
 */
